doc: Add Clang/LLVM based coverage report generation

Prabhat1308 opened the pull request:
Followed up from the [comment](https://github.com/bitcoin/bitcoin/issues/31927#issuecomment-2674522975) on the issue [#31927](https://github.com/bitcoin/bitcoin/issues/31927) , issues have been observed building coverage reports with `gcov` in MacOs and NixOs. This PR adds the steps to generate a coverage report based on the default llvm/clang tooling.

maflcko reviewed (approved):
lgtm

hebasto commented:
cc @vasild

Prabhat1308 commented on doc/developer-notes.md:
```diff
@@ -513,6 +515,46 @@ To enable test parallelism:
 
+#### Using LLVM/Clang tooling
+
+Default LLVM/Clang tooling can be used to generate coverage report for unit and functional tests.
```
updated.

Prabhat1308 commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,46 @@ To enable test parallelism:
+```
+
+The coverage report will be generated in the `coverage_report` directory inside the `build` directory.
 ### Performance profiling with perf
````
added the newline

jonatack commented on doc/developer-notes.md:
```diff
@@ -513,6 +515,47 @@ To enable test parallelism:
+    --show-line-counts-or-regions \
+    --show-expansions \
+    --output-dir=coverage_report \
+    --project-title="Bitcoin Core Coverage Report"
```
Following these instructions, this command failed for me with

`error: coverage.profdata: could not read profile data!No such file or directory`.

I then ran `rm -rf ./build/` and re-ran the instructions. Now seeing:

`warning: 2 functions have mismatched data`.

Haven't looked further yet.

I usually build with both `rm -rf ./build/` and `-DCMAKE_C_COMPILER="$(brew --prefix llvm)/bin/clang" -DCMAKE_CXX_COMPILER="$(brew --prefix llvm)/bin/clang++"`, among other flags, so I may need to try that.

Prabhat1308 replied:
The first error never occurs if I always remove my `build` folder and then proceeding with the commands specified. 

For MacOs , I have llvm 19 which is also a known [issue](https://github.com/bitcoin/bitcoin/issues/31049) so I generally have to specify my commands with this to use llvm 18.

```
cmake -B build \
   -DCMAKE_C_COMPILER="$(brew --prefix llvm@18)/bin/clang" \
   -DCMAKE_CXX_COMPILER="$(brew --prefix llvm@18)/bin/clang++" \
   -DCMAKE_C_FLAGS="-fprofile-instr-generate -fcoverage-mapping" \
   -DCMAKE_CXX_FLAGS="-fprofile-instr-generate -fcoverage-mapping" \
   -DAPPEND_LDFLAGS="-Wl,-no_warn_duplicate_libraries" \
```
and also replace
`llvm-profdata` with `/opt/homebrew/opt/llvm@18/bin/llvm-profdata` and `llvm-cov` with `/opt/homebrew/opt/llvm@18/bin/llvm-cov`

I do get the warning . The warning can be because of many reasons like Link-Time Optimisations , multiple defined compilations units etc. which results in different profile data. Its harmless and can be ignored.

pablomartin4btc replied:
> Its harmless and can be ignored.

I got it in both Ubuntu and macOS. Perhaps we can document it as well to avoid confusion?

```
$ find build -name "default_*.profraw" | xargs llvm-profdata merge -sparse -o build/coverage.profdata
build/src/univalue/default_44199.profraw: main: function basic block count change detected (counter mismatch)
Make sure that all profile data to be merged is generated from the same binary.
$ ll build/coverage.profdata
... 15926688 feb 27 09:44 build/coverage.profdata
$ llvm-cov show build/src/test/test_bitcoin \
    --instr-profile=build/coverage.profdata \
    --format=html \
    --show-instantiation-summary \
    --show-line-counts-or-regions \
    --show-expansions \
    --output-dir=build/coverage_report \
    --project-title="Bitcoin Core Coverage Report"
warning: 2 functions have mismatched data
```

Prabhat1308 replied:
> I got it in both Ubuntu and macOS. Perhaps we can document it as well to avoid confusion?
> 
> ```
> $ find build -name "default_*.profraw" | xargs llvm-profdata merge -sparse -o build/coverage.profdata
> build/src/univalue/default_44199.profraw: main: function basic block count change detected (counter mismatch)
> Make sure that all profile data to be merged is generated from the same binary.
> ```

Have you tried building with clang 19 ? I dont get this on MacOs
```
╰─ find build -name "default_*.profraw" | xargs llvm-profdata merge -sparse -o build/coverage.profdata


╭────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── ✔ 21:32:24 
 llvm-cov show build/src/test/test_bitcoin \                                                 
    --instr-profile=build/coverage.profdata \
    --format=html \
    --show-instantiation-summary \
    --show-line-counts-or-regions \
    --show-expansions \
    --output-dir=build/coverage_report \
    --project-title="Bitcoin Core Coverage Report"
warning: 2 functions have mismatched data
```

pablomartin4btc replied:
Ok, I had 18.1.8, updated it to 19, rerun the merge and I can confirm that the "counter mismatch" has gone. nit: Please consider adding a note there regarding the message and the warning with the 2 functions mismatch, any of those didn't affect the generation of the report nor its visualisation. Thanks!

Prabhat1308 replied:
Added the notes under the commands.

jonatack reviewed (commented):
Concept ACK

brunoerg reviewed (commented):
Concept ACK

hodlinator reviewed (commented):
Concept ACK 6436bdb301278abb3526eab2138b2f23f8905b5e

Prabhat1308 committed:
add clang/llvm based coverage report generation

Signed-off-by: Prabhat Verma <prabhatverma329@gmail.com>

Prabhat1308 commented:
Addressed the comments
1. Change the commands so we dont have to change shell from the root of the repo.
2. Added a comment for MacOs users to specify the command changes needed to use different llvm version if faced with issues.

hodlinator commented on doc/developer-notes.md:
```diff
@@ -513,6 +515,47 @@ To enable test parallelism:
 
+#### Using LLVM/Clang tooling
+
+Default LLVM/Clang tooling can be used to generate coverage report for unit tests.
```
I see. According to *doc/dependencies.md*, one only needs to have one of the compilers installed. I think it's clearer to remove the "Default" qualifier and needless to repeat the name of the toolchain.
```suggestion
The following generates a coverage report for unit tests.
```

hodlinator commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,45 @@ To enable test parallelism:
 cmake -DJOBS=$(nproc) -P build/Coverage.cmake
 ```
 
+#### Using LLVM/Clang tooling
````
nit: Could use more frequently used term:
```suggestion
#### Using LLVM/Clang toolchain
```

hodlinator commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,45 @@ To enable test parallelism:
+
+Configure the build with the following flags:
+
+```shell
+cmake -B build -DCMAKE_C_COMPILER="clang" \
+   -DCMAKE_CXX_COMPILER="clang++" \
+   -DCMAKE_C_FLAGS="-fprofile-instr-generate -fcoverage-mapping" \
+   -DCMAKE_CXX_FLAGS="-fprofile-instr-generate -fcoverage-mapping"
+  # for MacOS users , replace -DCMAKE_CXX_COMPILER="clang++" with -DCMAKE_CXX_COMPILER="$(brew --prefix llvm)/bin/clang++" and -DCMAKE_C_COMPILER="clang" with -DCMAKE_C_COMPILER="$(brew --prefix llvm)/bin/clang"
+cmake --build build # Use "-j N" here for N parallel jobs.
+```
````
No sure about the random 2-space indentation, also prefer comments preceding the command they're about.
````suggestion
```shell
# MacOS may require -DCMAKE_CXX_COMPILER="$(brew --prefix llvm)/bin/clang++" and -DCMAKE_C_COMPILER="$(brew --prefix llvm)/bin/clang"
cmake -B build -DCMAKE_C_COMPILER="clang" \
   -DCMAKE_CXX_COMPILER="clang++" \
   -DCMAKE_C_FLAGS="-fprofile-instr-generate -fcoverage-mapping" \
   -DCMAKE_CXX_FLAGS="-fprofile-instr-generate -fcoverage-mapping"
cmake --build build # Use "-j N" here for N parallel jobs.
```
````

hodlinator commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,45 @@ To enable test parallelism:
+    --project-title="Bitcoin Core Coverage Report"
+```
+
+The coverage report will be generated in the `coverage_report` directory inside the `build` directory.
````
nit: A bit verbose to repeat "directory", more helpful to include *index.html*?
```suggestion
The generated coverage report can be accessed at `build/coverage_report/index.html`.
```

hodlinator reviewed (commented):
Reviewed 5f7b42642ad83faf23c8e75511f2379f9eae25ad

- Re-testing the commands.
- Managed to generate coverage for a functional tests, not sure if we should include it, includes copy-pasting temporary paths for the root test directory.

janb84 commented:
Concept ACK [5f7b426](https://github.com/bitcoin/bitcoin/pull/31933/commits/5f7b42642ad83faf23c8e75511f2379f9eae25ad)

Using the new instructions of the  document on MacOS using Nix-shell (so using the non mac instructions) I do endup gettting the warning that was mentioned before: 
**"warning: 2 functions have mismatched data"**

The rapport is still generated.  

Tested both under clang 19.1.7 / 18.1.8 and LLVM 19.1.7 / 18.1.8

This is an improvement because I cannot get the "old way" of coverage generation to work!

janb84 reviewed (commented):
Concept ACK [5f7b426](https://github.com/bitcoin/bitcoin/pull/31933/commits/5f7b42642ad83faf23c8e75511f2379f9eae25ad)

Using the new instructions of the document on MacOS using Nix-shell (so using the non mac instructions) I do endup gettting the warning that was mentioned before:
"warning: 2 functions have mismatched data"

The rapport is still generated.

Tested both under clang 19.1.7 / 18.1.8 and LLVM 19.1.7 / 18.1.8

This is an improvement because I cannot get the "old way" of coverage generation to work!

Prabhat1308 commented:
> * Managed to generate coverage for a functional tests, not sure if we should include it, includes copy-pasting temporary paths for the root test directory.

I think adding the coverage for functional tests and fuzz tests (which I believe everybody does using llvm toolchain only ?) can be added in another PR .

janb84 commented:
re ACK [05605da](https://github.com/bitcoin/bitcoin/commit/05605dae2983c11f0ba662dbdc495f84de12724b)

pablomartin4btc reviewed (approved):
tACK 05605dae2983c11f0ba662dbdc495f84de12724b

<details>
<summary>Tested in both Ubuntu 22.04 (with clang 16) and macOS 14.4</summary>

![image](https://github.com/user-attachments/assets/69fadb98-6ff7-4bf1-8180-fd29a2f03547)

![image](https://github.com/user-attachments/assets/a8234f92-0cf0-4fbd-91fe-2457f317801a)

</details>

Maybe we could advice the user about the warnings, left a comment there.

BrandonOdiwuor reviewed (commented):
Concept ACK

hodlinator commented on doc/developer-notes.md:
```diff
@@ -484,6 +484,8 @@ $ ./build/test/functional/test_runner.py --valgrind
 
```
I just realized the PR description shouldn't contain the @-symbol. See:

https://github.com/bitcoin/bitcoin/blob/3c1f72a36700271c7c1293383549c3be29f28edb/CONTRIBUTING.md?plain=1#L170-L174

Prabhat1308 replied:
Removed the @ symbol

hodlinator reviewed (approved):
ACK 1245e05325b41101eddc76ba9214d910489e35b6

- Good to include some comments on expected warnings (I was getting some too).
- Thanks for incorporating my latest feedback!
- Agree that functional/fuzz test instructions could either be a separate PR or left as an exercise to the reader.

hodlinator commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,49 @@ To enable test parallelism:
 cmake -DJOBS=$(nproc) -P build/Coverage.cmake
 ```
 
````
nit: Could shorten PR title. (Doesn't invalidate ACKs).
```diff
-doc: Update documentation to include Clang/llvm based coverage report generation
+doc: Add Clang/LLVM based coverage report generation
```

janb84 commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,49 @@ To enable test parallelism:
+LLVM_PROFILE_FILE="default_%p.profraw" ctest --test-dir build # Use "-j N" here for N parallel jobs.
+
+# merge all the raw profile data into a single file
+find build -name "default_*.profraw" | xargs llvm-profdata merge -sparse -o build/coverage.profdata
+# Note: The "counter mismatch" warning can be safely ignored, though it can be resolved by updating to Clang 19.
+# The warning occurs due to version mismatches but doesn't affect the coverage report generation.
+```
+
````
Small nit, creating a blockquote note remark section is maybe more fitting:

> Note: The "counter mismatch" warning can be safely ignored, though it can be resolved by updating to Clang 19.\
> The warning occurs due to version mismatches but doesn't affect the coverage report generation.

janb84 reviewed (commented):
re ACK [1245e05](https://github.com/bitcoin/bitcoin/commit/1245e05325b41101eddc76ba9214d910489e35b6)

The extra comments are informative of the possible warnings and their influence on the result.

vasild commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,49 @@ To enable test parallelism:
+
+```shell
+LLVM_PROFILE_FILE="default_%p.profraw" ctest --test-dir build # Use "-j N" here for N parallel jobs.
+
````
I would suggest collecting coverage from functional tests as well because that would change a bit the arguments of `llvm-cov` (see later) and because the path to the profile file has to be specified as an absolute path which is not obvious:

```md
Generating the raw profile data based on `ctest` execution:

```shell
LLVM_PROFILE_FILE=$(pwd)/build/raw_profile_data/%m_%p.profraw build/test/functional/test_runner.py
`` `
```
it has to be a full path because otherwise the `.profraw` file will be created inside the temporary directory, like `/tmp/test_runner_₿_🏃_20250304_134152/...` and will be removed when the temporary directory is deleted at the end of the tests.

vasild commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,49 @@ To enable test parallelism:
+Generating the coverage report:
+
+```shell
+llvm-cov show build/src/test/test_bitcoin \
+    --instr-profile=build/coverage.profdata \
+    --format=html \
+    --show-instantiation-summary \
+    --show-line-counts-or-regions \
+    --show-expansions \
+    --output-dir=build/coverage_report \
+    --project-title="Bitcoin Core Coverage Report"
````
If there are profile files from more than one executable, like from `test_bitcoin` and from `bitcoind` then the syntax of this changes a little bit: drop `build/src/test/test_bitcoin` and add `-object=build/src/test/test_bitcoin -object=build/src/bitcoind`:

```
llvm-cov show 
    --object=build/src/test/test_bitcoin \
    --object=build/src/bitcoind \
    --instr-profile=build/coverage.profdata \
    --format=html \
    --show-instantiation-summary \
    --show-line-counts-or-regions \
    --show-expansions \
    --output-dir=build/coverage_report \
    --project-title="Bitcoin Core Coverage Report"
```

vasild commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,49 @@ To enable test parallelism:
+Generating the coverage report:
+
+```shell
+llvm-cov show build/src/test/test_bitcoin \
+    --instr-profile=build/coverage.profdata \
````
Locally, I have been using this as well:

```
-ignore-filename-regex="src/crc32c/|src/leveldb/|src/minisketch/|src/secp256k1/|src/test/"
```

vasild reviewed (commented):
Approach ACK 1245e05325b41101eddc76ba9214d910489e35b6

In the light of https://github.com/bitcoin/bitcoin/issues/31047 and https://github.com/bitcoin/bitcoin/pull/31394 I have been thinking that it would be useful to at least document the steps needed to gather coverage with clang. Thanks for the initiative!

Prabhat1308 commented:
Addressing the review from [comment](https://github.com/bitcoin/bitcoin/pull/31933#pullrequestreview-2656773286)
- The coverage now includes functional tests 
- Removed coverage from external deps (crc32,leveldb,minisketch,secp256k1) and `src/test`.
- Now uses `-DCMAKE_BUILD_TYPE=Debug`
- Notes are now in Blockquotes
-  Incuding the functional tests has increased the mismatch from 2 to 61 .

pablomartin4btc commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,60 @@ To enable test parallelism:
+cmake --build build # Use "-j N" here for N parallel jobs.
+```
+
+Generating the raw profile data based on `ctest` execution:
````
small nit:
```suggestion
Generating the raw profile data based on `ctest` and functional tests execution:
```
Or remove both and clarify it below before each execution (or any alternative you might have).

pablomartin4btc reviewed (commented):
ACK 7f57ad00d301b9564ece511d2e99cb0678b21174

I think the feedback about [`CMAKE_BUILD_TYPE`](https://github.com/bitcoin/bitcoin/pull/31933#discussion_r1993037907) needs to be addressed.

Since the functional tests were added (for the report generation), is it worth adding the `fuzz`?

vasild commented on doc/developer-notes.md:
```diff
@@ -513,6 +515,49 @@ To enable test parallelism:
+cmake -B build -DCMAKE_C_COMPILER="clang" \
+   -DCMAKE_CXX_COMPILER="clang++" \
+   -DCMAKE_C_FLAGS="-fprofile-instr-generate -fcoverage-mapping" \
+   -DCMAKE_CXX_FLAGS="-fprofile-instr-generate -fcoverage-mapping"
```
@ldionne, do you know if `-O2` could have any adverse effects on coverage created with clang's `-fprofile-instr-generate -fcoverage-mapping`?

ldionne replied:
Sadly, I am not very familiar with how code coverage works in Clang (or other compilers), so I am not the best person to answer that question. However, as a general rule of thumb, I would encourage you to go with what the documentation says, and if you then notice shortcomings, report those as bugs against the project.

vasild replied:
The docs at https://clang.llvm.org/docs/SourceBasedCodeCoverage.html don't mention anything about `-O` flags.

vasild replied:
I don't see how from "source-based ... operates on AST and preprocessor information" follows that `-O` don't matter.

I did coverage reports locally with `Debug` (`-O0`) and `Release` (`-O2`) and compared them. They are slightly different, which could be due to non-determinism in the tests. `src/sync.cpp` is completely missing from the `-O2` report.

Time to complete the tests:

test | Debug | Release
--- | --- | ---
unit | 60s | 31s
functional | 186s | 76s

ldionne replied:
The `-O` flag controls optimization level, which shouldn't have any impact on the AST representation of the code. The compiler builds an AST representation of the code that corresponds to the semantics of the program, and from that generates LLVM IR that implements those semantics. This is then passed on to the optimizer, which performs transformations on the LLVM IR representation (not the AST directly) to produce equivalent but more efficient LLVM IR.

This is much simplified, but if code coverage is implemented based on the AST representation, then I do agree that `-O` flags shouldn't matter. That being said, I have a hard time imagining how each line in the final IR can always be attributed to the correct line in the original sources in the presence of optimizations like inlining and especially outlining, where the same executable code may be used to implement source-different functions. But if the documentation says it, I would go with that.

vasild replied:
@ldionne, that is very useful. Thanks for the clarification! With my limited knowledge I was thinking that maybe AST is generated after optimization.

Just to clarify on this particular issue - I am fine with whatever `CMAKE_BUILD_TYPE=` or omitting it and leaving the default. This is only a documentation after all. People can tweak the commands as they wish before executing them.

vasild commented on doc/developer-notes.md:
```diff
@@ -513,6 +515,60 @@ To enable test parallelism:
+
+# Run tests to generate profiles
+LLVM_PROFILE_FILE="$(pwd)/build/raw_profile_data/%m_%p.profraw" ctest --test-dir build # Use "-j N" here for N parallel jobs.
+LLVM_PROFILE_FILE="$(pwd)/build/raw_profile_data/%m_%p.profraw" build/test/functional/test_runner.py
```
nit: `test_runner.py` also supports `-j N` for parallel jobs.

vasild commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,60 @@ To enable test parallelism:
+
+```shell
+llvm-cov show \
+    --object=build/src/test/test_bitcoin \
+    --object=build/src/bitcoind \
````
After https://github.com/bitcoin/bitcoin/pull/31161 the binaries are now in `bin/`:

```suggestion
    --object=build/bin/test_bitcoin \
    --object=build/bin/bitcoind \
```

vasild reviewed (approved):
Almost ACK 7f57ad00d301b9564ece511d2e99cb0678b21174, just fix the path to the binaries which were recently relocated from `src` to `bin`.

I would be happy to review an extension to this that also does the fuzz part. That would involve compiling into another directory for fuzzing, running the fuzz tests, gathering their `.profraw` files as well and adding `--object=buildfuzz/bin/fuzz` to `llvm-cov`.

Prabhat1308 commented:
Addressed the comments to fix nits. On the use of `-DEBUG` , I haven't been able to do a comparison from lcov generated reports as I don't have any other system . Since `-02` introduces optimisation , I have chosen to keep the `-DEBUG` flag to use `-00` as mentioned here to keep it close to gcov reports.
https://stackoverflow.com/questions/36930207/code-coverage-with-optimization
> it seems if I want to collect coverage information with tools like gcov, any optimization flag must be disabled

janb84 commented on doc/developer-notes.md:
```diff
@@ -513,6 +515,60 @@ To enable test parallelism:
+# MacOS may require -DCMAKE_CXX_COMPILER="$(brew --prefix llvm)/bin/clang++" and -DCMAKE_C_COMPILER="$(brew --prefix llvm)/bin/clang"
+cmake -B build -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_COMPILER="clang" \
+   -DCMAKE_CXX_COMPILER="clang++" \
+   -DAPPEND_CFLAGS="-fprofile-instr-generate -fcoverage-mapping" \
+   -DAPPEND_CXXFLAGS="-fprofile-instr-generate -fcoverage-mapping" \
```
```suggestion
   -DAPPEND_CFLAGS="-fprofile-instr-generate -fcoverage-mapping -g" \
   -DAPPEND_CXXFLAGS="-fprofile-instr-generate -fcoverage-mapping -g" \
```
NIT: In Debug builds, the compiler usually disables optimizations, but without explicit -g, some optimizations might still occur. Given that the code coverage results are indeed different when running without `-g` and with `-g`, I would say some optimization still occurs interfering with the code coverage results.

janb84 replied:
> `-DCMAKE_BUILD_TYPE=Debug` adds `-g` already.

Why is there a difference in the reports when manually adding the -g ?

janb84 replied:
The extra `-g` just gives a lot of `Unexecuted instantiation:` messages and a lot lower coverage. No sure why but don't think this gives a good image of the actual coverage.

janb84 reviewed (commented):
~~While re-testing this PR, I found that adding the `-g` flag (debug symbols) impacts code coverage reports. With `-g`, more source lines are mapped, lowering the coverage percentage by revealing untested code that’s hidden without it. I tested this with and without `-DCMAKE_BUILD_TYPE=Debug`, and both setups give a different result.~~
Disproven it gives a lot of `Unexecuted instantiation:` errors.


I agree with the rest of the PR changes.  ACK [3c69b07](https://github.com/bitcoin/bitcoin/commit/3c69b071ac4964870abb347f9d9c15df2e60aa62)

vasild reviewed (approved):
ACK 3c69b071ac4964870abb347f9d9c15df2e60aa62

Thank you!

jonatack commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,60 @@ To enable test parallelism:
+
+```shell
+# MacOS may require -DCMAKE_CXX_COMPILER="$(brew --prefix llvm)/bin/clang++" and -DCMAKE_C_COMPILER="$(brew --prefix llvm)/bin/clang"
+cmake -B build -DCMAKE_BUILD_TYPE=Debug -DCMAKE_C_COMPILER="clang" \
````
Directory name: would it make sense to replace the `build` directory in the instructions with, say, `build_cov` instead?

I think we're already using directory naming like `build_fuzz` (in `doc/fuzzing.md`) and `build_dev_mode` (in `CMakePresets.json`).

Prabhat1308 replied:
Decided against it since the previous docs built the coverage in `build` directory .

jonatack commented on doc/developer-notes.md:
```diff
@@ -513,6 +515,60 @@ To enable test parallelism:
+> **Note:** The "functions have mismatched data" warning can be safely ignored, the coverage report will still be generated correctly despite this warning.
+> This warning occurs due to profdata mismatch created during the merge process for shared libraries.
+
+The generated coverage report can be accessed at `build/coverage_report/index.html`.
```
```suggestion
The generated coverage report is located at `build/coverage_report/index.html`. Running `open build/coverage_report/index.html` will open it in your browser.
```

hodlinator replied:
Thread https://github.com/bitcoin/bitcoin/pull/31933#discussion_r1999113012:

`open` is MacOS-specific AFAIK. I don't have it on my Linux install. Also overly repetitive/verbose IMO.

jonatack commented on doc/developer-notes.md:
```diff
@@ -513,6 +515,60 @@ To enable test parallelism:
+
+The following generates a coverage report for unit tests and functional tests.
+
+Configure the build with the following flags:
```
Should we mention clearing out with `rm -rf build` to begin with a clean slate?

jonatack reviewed (commented):
ACK 3c69b071ac4964870abb347f9d9c15df2e60aa62 modulo feedback below

vasild commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,60 @@ To enable test parallelism:
+Generating the coverage report:
+
+```shell
+llvm-cov show \
+    --object=build/bin/test_bitcoin \
````
I forgot to mention one more option that is useful here: `-Xdemangler=llvm-cxxfilt`. Without that option I get:
```
Unexecuted instantiation: addrman_tests.cpp:_ZL17LogAcceptCategoryN5BCLog8LogFlagsENS_5LevelE
```
and with that option:
```
Unexecuted instantiation: addrman_tests.cpp:LogAcceptCategory(BCLog::LogFlags, BCLog::Level)
```

ryanofsky reviewed (commented):
@Prabhat1308 this seems ready to merge but also has a few recent comments that haven't been responded to. You can let me know if you'd like this to be merged or if you want to make more changes.

Prabhat1308 commented:
- Following the discussion , I have removed the `-DEBUG` flag .
-  Added `-Xdemangler=llvm-cxxfilt` option for more readable outputs.
- Added the note to consider with a clean state to avoid these issues https://github.com/bitcoin/bitcoin/pull/31933#discussion_r1966580329
- Added some suggestions by @jonatack

Prabhat1308 commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,60 @@ To enable test parallelism:
+Configure the build with the following flags:
+
+```shell
+# MacOS may require -DCMAKE_CXX_COMPILER="$(brew --prefix llvm)/bin/clang++" and -DCMAKE_C_COMPILER="$(brew --prefix llvm)/bin/clang"
````
Added , although I prefer it being before the commands then after it.

hodlinator replied:
Thread https://github.com/bitcoin/bitcoin/pull/31933#discussion_r1999092770:

nit: Would prefer not repeating the text below, and think mentioning brew is overly verbose, adding "instead" is sufficient IMO:
```
# MacOS may instead require `-DCMAKE_C_COMPILER="$(brew --prefix llvm)/bin/clang" -DCMAKE_CXX_COMPILER="$(brew --prefix llvm)/bin/clang++"`
```

Prabhat1308 commented:
@ryanofsky I have addressed the remaining comments and these should be my last changes (if everyone agrees)

hodlinator commented on doc/developer-notes.md:
```diff
@@ -513,6 +515,49 @@ To enable test parallelism:
+LLVM_PROFILE_FILE="default_%p.profraw" ctest --test-dir build # Use "-j N" here for N parallel jobs.
+
+# merge all the raw profile data into a single file
+find build -name "default_*.profraw" | xargs llvm-profdata merge -sparse -o build/coverage.profdata
```
Thread https://github.com/bitcoin/bitcoin/pull/31933#discussion_r1979323892:

With `-sparse`:
```
warning: 16 functions have mismatched data
```
Without `-sparse` (current version of PR):
```
warning: 65 functions have mismatched data
```

hodlinator commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,64 @@ To enable test parallelism:
+
+```shell
+# Create directory for raw profile data
+mkdir -p $(pwd)/build/raw_profile_data
````
`$(pwd)` is redundant for `mkdir`:
```suggestion
mkdir -p build/raw_profile_data
```

Prabhat1308 replied:
done.

hodlinator commented on doc/developer-notes.md:
```diff
@@ -513,6 +515,64 @@ To enable test parallelism:
+LLVM_PROFILE_FILE="$(pwd)/build/raw_profile_data/%m_%p.profraw" ctest --test-dir build # Use "-j N" here for N parallel jobs.
+LLVM_PROFILE_FILE="$(pwd)/build/raw_profile_data/%m_%p.profraw" build/test/functional/test_runner.py # Use "-j N" here for N parallel jobs
+
+# merge all the raw profile data into a single file
+find build -name "*.profraw" | xargs llvm-profdata merge -o build/coverage.profdata
```
nit: Could be more precise, use leading Capital letter like other comments.
```suggestion
# Merge all the raw profile data into a single file
find build/raw_profile_data -name "*.profraw" | xargs llvm-profdata merge -o build/coverage.profdata
```

hodlinator reviewed (commented):
Reviewed/tested 3b5d00f65ae37b733db74d9c9c905eb35e7363e5

Prabhat1308 commented:
Addressed the reviews by @hodlinator

hodlinator reviewed (approved):
re-ACK b96f1a696aa792068b5a4fa16e2d4a342e4f55b8

Thanks for addressing my feedback!

Still think it's strange that removing `-sparse` increases the functions with [mismatched data](https://github.com/bitcoin/bitcoin/pull/31933#discussion_r2013585005). But it's certainly not blocking for me.

janb84 reviewed (commented):
Re ACK [b96f1a6](https://github.com/bitcoin/bitcoin/commit/b96f1a696aa792068b5a4fa16e2d4a342e4f55b8)

After reading the discussion and the Clang documentation, I think that the call to remove the DEBUG build type is a good call. Agree with other changes. 
Good work !

fanquake commented on doc/developer-notes.md:
```diff
@@ -513,6 +515,63 @@ To enable test parallelism:
+
+Configure the build with the following flags:
+
+> Consider building with a clean state using `rm -rf build`
```
We don't do this in any other docs, and as far as I'm aware, this being needed would be a bug, not sure it should be here.

fanquake replied:
> but if it is important to remove, it could be done in a follow-up.

The more I look at this, the less clear it seems to be. Clearly the dir reuse can/is causing issues, however the discussion in #31942 also claims that any need to nuke a build dir is a bug. If we are going to start adding this, we should at least be consistent, and add this same disclaimer to every bit of documentation that invokes cmake. Will followup in #31942.

Prabhat1308 commented:
Don't know how the runs got cancelled.

Prabhat1308 commented:
@ryanofsky sorry for another ping. All the comments have been addressed and there is no other objection . The PR is ready to be merged.

Prabhat1308 commented:
@jonatack @pablomartin4btc @vasild would appreciate some re-ACKs.

i-am-yuvi commented:
Concept ACK

Crypt-iQ commented:
tACK b96f1a696aa792068b5a4fa16e2d4a342e4f55b8

I was able to produce coverage reports with `llvm@18` on macOS 15.4.  I wasn't able to produce a coverage report for the fuzzer following similar steps -- the `raw_profile_data` directory was empty.

Prabhat1308 commented:
> tACK [b96f1a6](https://github.com/bitcoin/bitcoin/commit/b96f1a696aa792068b5a4fa16e2d4a342e4f55b8)
> 
> I was able to produce coverage reports with `llvm@18` on macOS 15.4. I wasn't able to produce a coverage report for the fuzzer following similar steps -- the `raw_profile_data` directory was empty.

The following steps are only for functional tests and unit tests. I will add coverage generation for fuzz tests in another follow up PR .

Prabhat1308 commented on doc/developer-notes.md:
````diff
@@ -513,6 +515,63 @@ To enable test parallelism:
+Generating the coverage report:
+
+```shell
+llvm-cov show \
+    --object=build/bin/test_bitcoin \
+    --object=build/bin/bitcoind \
````
There are also other binaries like `bitcoin-cli` , `bitcoin-tx` including `bitcoin-chainstate` which are built when using specific flags. Including each requires `--object={}` . Adding too many felt too verbose and not needed by a general reviewer. If required he/she may add it themselves.

vasild commented:
ACK b96f1a696aa792068b5a4fa16e2d4a342e4f55b8
